driver: host_uart: Add host-uart device in npcx.

MulinChao opened the pull request:
In order to prevent user turns on the pin-mux of devices has io-pads unexpectedly, this CL added a new device definition for host uart device. The pin-mux of host uart interface is enabled only if we set its status as "okay" in dts file of board folder.

The following npcx7 drivers will meet:
1. Default status property of all npcx devices with io-pads such as espi, pwm, uart, host uart and so on should be "disabled".
2. Switch pin-mux by changing status property to "okay" in dts file of board folder.
3. Add necessary definitions in prj file of board or the other application folder.

Signed-off-by: Mulin Chao <MLChao@nuvoton.com>

jettr commented on dts/arm/nuvoton/npcx7m6fb.dtsi:
```diff
@@ -453,12 +453,6 @@
 					  "pmch_obe", "p80_fifo";
 
-			/* Host serial port pinmux PIN75 86 36 33 42 C7 B3 B2 */
-			pinctrl = <&altb_rxd_sl &altb_txd_sl
```
Are these these the pins needed for LPC/eSPI communication? If so can we leave them here in the and make the status of this device "disabled" instead?

Would it be possible to enable the LPC device and leave the io_host_uart disabled? When would that scenario exist?

MulinChao replied:
> Are these these the pins needed for LPC/eSPI communication? If so can we leave them here in the and make the status of this device "disabled" instead?
[ML] These pins are independent of LPC/eSPI communication. eSPI works normally if pin-mux of host uart is off.
> Would it be possible to enable the LPC device and leave the io_host_uart disabled? When would that scenario exist?
[ML] As far as I know, there is only one status property in one device with the same compatible. 

The following is my original plan for pin-mux. Any comment is appreciated.
First, the "status" of internal modules without io-pads such as pcc, scfg, host_sub, and so on are "okay" by default. You need to turn them on by enabling CONFIG_CLOCK_CONTROL, CONFIG_PINMUX, and CONFIG_ESPI in defconfig or prj files.

Then, the "status" of devices with io-pads (with "pinctrl" property) besides GPIO must be "disabled" by default. Besides turning their configuration, you also need to set their status to "okay" in dts file explicitly. For example, the following steps are necessary if we need host uart functionality in our application.

1. Add CONFIG_ESPI=y and CONFIG_ESPI_PERIPHERAL_UART=y in prj file of samples folder.
2. Add &host_uart{ status = "okay";}; in dts file.

The benefit of this method is easy to observe the pin-mux usages of all io-pads in npcx ec in dts file. By this simple rule, we needn't to spend time to check the actual pin-mux usage in our driver.

jettr replied:
Sounds good. Thanks for the explanation.

MulinChao commented:
Hi all: 
This PR is pending for a while? Any comment is appreciated.

MaureenHelm reviewed (commented):
@galak please review the pinmuxing

MulinChao commented:
Hi all:
Regarding the rename of 'pinctrl' property, I have collected all the changes in the [29384](https://github.com/zephyrproject-rtos/zephyr/pull/29384). Since it's also related to macro functions here, I plan to do rebase after this PR, [PR29144](https://github.com/zephyrproject-rtos/zephyr/pull/29144), and [PR29035](https://github.com/zephyrproject-rtos/zephyr/pull/29035) are landed. Hope it is clearer to review.

galak commented on dts/arm/nuvoton/npcx7m6fb.dtsi:
```diff
@@ -470,6 +464,17 @@
+		host_uart: io_host_uart {
+			compatible = "nuvoton,npcx-host-uart";
+			/* Host serial port pinmux PIN75 86 36 33 42 C7 B3 B2 */
+			pinctrl = <&altb_rxd_sl &altb_txd_sl
```
probably needs updating to pinctrlr-0

MulinChao replied:
Done.

galak commented on dts/bindings/espi/nuvoton,npcx-host-uart.yaml:
```diff
@@ -0,0 +1,14 @@
+include: [base.yaml]
+
+properties:
+    pinctrl:
```
rename to pinctrl-0

MulinChao replied:
Done.

galak reviewed (changes requested):
need to rename pinctrl property to pinctrl-0

MulinChao commented:
> need to rename pinctrl property to pinctrl-0

Since PR29384 has been merged first, I will rebase and change this property name later.

MulinChao commented:
Hi @galak: 
Rename stuff is done. Please review it. Thanks.

MulinChao commented:
> need to rename pinctrl property to pinctrl-0

Hi @galak:
I have finished the changes of renaming from pinctrl property to pinctrl-0. Any comment is appreciated. Thanks.

jackrosenthal commented:
@galak can you mark your changes requested as complete so we can merge?

MulinChao commented:
Hi @galak, @jettr, and @keithshort:
The changes for request has been submitted for a while, Might I know what I can do for merging? Thanks.

jettr commented:
@nashif Can you help us figure out next steps to get this merged, please?

jettr commented:
@MulinChao please remove @galak  from the review. I believe that his change request is blocking the automatic merge process (and you have address his feedback)

nashif commented:
@galak can you please take another look?

galak commented on soc/arm/nuvoton_npcx/common/soc_dt.h:
```diff
@@ -166,6 +157,114 @@
+ * @return a node identifier for the node with "io_comp" compatible and
+ *         instance number "inst"
+ */
+#define DT_NPCX_COMP_INST(compat, inst) DT_INST(inst, compat)
```
Please don't use `DT_` prefix for such macros, we try and reserve the name space for system DT macros.

MulinChao replied:
I checked all macros in "soc_dt.h" and modified the macros which prefix doesn't meet DT_NPCX that follows ATMEL did here(https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/arm/atmel/pinctrl_atmel_sam.h#L39). Is it OK with you?
Or is NPCX_DT prefix better?

pabigot replied:
I think the atmel macros are intended for use in devicetree sources, as they generate pinmux data, so they should have a `DT_` prefix.

       soc {
                pinctrl@400e0e00 {
                        /* instance, signal, pio, pin, peripheral */
                        DT_ATMEL_PIN(spi0, miso, a, 25, a);
                        DT_ATMEL_PIN(spi0, mosi, a, 26, a);
                        ...

These macros instead are used in C code to extract information from translated devicetree sources.  They *probably* should be `NPCX_DT_` since they don't live in `include/devicetree/*` where other similar system-wide macros are placed.

But @galak can provide authoritative direction.

MulinChao replied:
Re @pabigot:
Thanks. it's clear to me. I will replace all macros for constructing with NPCX_DT_ prefix. to prevent indistinct usage.

galak reviewed (dismissed):
The DT_ macros in soc_dt.h need to be renamed to not be `DT_` prefixed, as we reserve that prefix for system DT macros/defines.

MulinChao committed:
soc: replace DT_ with NPCX_DT_ prefix for all macros in soc_dt.h

This CL replaces all DT_ prefix with NPCX_DT_ for all macros used
for providing npcx device information in soc_dt.h It avoided the
ambiguity with the DT_ prefix for system DT macros/defines.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

MulinChao committed:
driver: host_uart: Add host-uart device in npcx.

In order to prevent user turns on the pin-mux of devices has io-pads
unexpectedly, this CL added a new device definition for host uart
device. The pin-mux of host uart interface is enabled only if we set its
status as "okay" in dts file of board folder.

The following npcx7 drivers will meet:
1. Default status property of npcx devices with io-pads such as espi,
   pwm, uart, host uart and so on should be "disabled".
2. Switch pin-mux by changing status property to "okay" in dts file of
   board folder.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

MulinChao commented:
> The DT_ macros in soc_dt.h need to be renamed to not be `DT_` prefixed, as we reserve that prefix for system DT macros/defines.
Hi @galak:
The change for the request is submitted. If you have further comments, please let me know. Thanks.

MulinChao commented:
Hi @galak & @pabigot:
I have replaced all macros which prefix are DT_ or DT_NPCX_ with NPCX_DT. Please take a look and any comment is appreciated. Thanks.

pabigot reviewed (approved):
Looks reasonable to me.

MulinChao commented:
Hi All:
Finally, hope this PR will be landed soon. Thanks for your comments. I really appreciate that.
